src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -167,15 +167,37 @@ private async Task<FormulaValue> TryHandleSet(CallNode node, EvalVisitorContext
 
             var newValue = await arg1.Accept(this, context).ConfigureAwait(false);
 
-            if (arg0.IRContext.IsMutation && arg0 is RecordFieldAccessNode rfan)
+            if (arg0.IRContext.IsMutation)
             {
-                var arg0value = await rfan.From.Accept(this, context).ConfigureAwait(false);
-                
-                if (arg0value is RecordValue rv)
+                if (arg0 is RecordFieldAccessNode rfan)
                 {
-                    rv.ShallowCopyFieldInPlace(rfan.Field);
-                    rv.UpdateField(rfan.Field, newValue);
-                    return node.IRContext.ResultType._type.Kind == DKind.Boolean ? FormulaValue.New(true) : FormulaValue.NewVoid();
+                    var arg0value = await rfan.From.Accept(this, context).ConfigureAwait(false);
+
+                    if (arg0value is RecordValue rv)
+                    {
+                        rv.ShallowCopyFieldInPlace(rfan.Field);
+                        rv.UpdateField(rfan.Field, newValue);
+                        return node.IRContext.ResultType._type.Kind == DKind.Boolean ? FormulaValue.New(true) : FormulaValue.NewVoid();
+                    }
+                    else
+                    {
+                        return CommonErrors.UnreachableCodeError(node.IRContext);
+                    }
+                }
+                else if (arg0 is BinaryOpNode bon && bon.Op == BinaryOpKind.DynamicGetField)
+                {
+                    var arg0value = await bon.Left.Accept(this, context).ConfigureAwait(false);
+
+                    if (arg0value is UntypedObjectValue uov)
+                    {
+                        TextLiteralNode textLiteralNode = (TextLiteralNode)bon.Right;

```

Comment:
>(TextLiteralNode) [](http://example.com/codeflow?start=58&length=17)

How do we know this cast is safe? I guess it's implied by DynamicGetField check on line 187 

Nit - consider sharing  a static helper with 
OperatorDynamicGetField

That can also capture this safety check.

Reply:
This is where Fx creates a `BinaryOpNode` and adds a `TextLiteralNode` to the right.
https://github.com/microsoft/Power-Fx/blob/b762be3ae8a45471ececad2d5b1b139361611b71/src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs#L787

Reply:
bool IsDynamicSetProperty(arg0, out UntypedObjectValue uo, out string propName)